hydroshare_jupyter_sync/lib/filesystem/fs_resource_map.py

```diff
@@ -0,0 +1,199 @@
+from abc import ABC, abstractmethod
+from collections import UserDict
+from typing import Dict, List, Union
+from hsclient import Resource
+import hashlib
+from pathlib import Path
+
+from .types import MD5Hash
+
+
+def compute_file_md5_hexdigest(file: Union[str, Path]) -> str:
+    """Compute a file's md5 hexdigest. Read file as chunks to conserve memory usage."""
+    with open(file, "rb") as f:
+        hash = hashlib.md5()
+        chunk = f.read(8192)
+        while chunk:
+            hash.update(chunk)
+            chunk = f.read(8192)
+
+        return hash.hexdigest()
+
+
+# abstract interfaces
+
+
+class IFSResourceMap(ABC):
+    """Class representing the relationship between a file (no directory) path to the file's MD5 Hash."""
+
+    @abstractmethod
+    def update_resource(self) -> None:
+        """Update entire resource map. md5hash's updated, non-existent files removed, and new files inserted."""
+
+
+class IEntityFSResourceMap(ABC):
+    """Class representing the relationship between a file (no directory) path to the file's MD5 Hash."""
+
+    @abstractmethod
+    def add_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Add file to FSResource map."""
+
+    @abstractmethod
+    def delete_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Remove file from FSResource map."""
+
+    @abstractmethod
+    def update_file(self, relative_resource_file: Union[Path, str]) -> None:
+        """Update file in FSResource map."""
+
+
+# semi-concrete implementations
+
+
+class FSResourceMap(UserDict, IFSResourceMap):
+    @property
+    def files(self) -> List[Path]:
+        """Return list of files in resource."""
+        return list(self.data.keys())
+
+
+# concrete implementations
+
+
+class LocalFSResourceMap(FSResourceMap, IEntityFSResourceMap):
+    """Concrete class representing the relationship between a local file (not directory) path to the file's MD5 Hash."""
+
+    def __init__(self, resource_path: Union[Path, str]) -> Dict[Path, MD5Hash]:
+        super().__init__()
+        self.resource_path = Path(resource_path).expanduser().resolve()
+        self.resource_id = resource_path.name
```

Comment:
This is probably fine with this application but it would be more thorough to check for resource id in the resourcemetadata.xml file.  What currently happens if the same resource is downloaded and unzipped twice, resulting in two copies of the resource and the most recently unzipped bag having a counter appended to the resource id folder name?

Reply:
That's a good point. Both this solution and your proposed have their own assumptions. Currently, if you download a single file from hydroshare it does not come with the `resourcemetadata.xml` file so that was the main driver for why I chose this route. Currently, the `LocalFSResourceMap` object does not care if the resource id is "correct." The only assumption that it makes is that there is a nested file structure, such that, `<resource-id>/<resource-id>/` exists. So the following would be invalid in the current implementation `0296bacd656b40308a2298174aecb9a6_1/0296bacd656b40308a2298174aecb9a6`.